server/athenian/api/models/persistentdata/versions/860cdc895ef3_add_release_notifications.py

```diff
@@ -0,0 +1,52 @@
+"""Add release notifications
+
+Revision ID: 860cdc895ef3
+Revises: 8dac2ad15a8d
+Create Date: 2021-02-19 06:36:10.938726+00:00
+
+"""
+from alembic import op
+import sqlalchemy as sa
+
+
+# revision identifiers, used by Alembic.
+revision = "860cdc895ef3"
+down_revision = "8dac2ad15a8d"
+branch_labels = None
+depends_on = None
+
+
+def upgrade():
+    name = "release_notifications"
+    if op.get_bind().dialect.name == "postgresql":
+        op.execute("CREATE SCHEMA athenian;")
+        schema_arg = {"schema": "athenian"}
+    else:
+        name = "athenian." + name
+        schema_arg = {}
+    op.create_table(
+        name,
+        sa.Column("account_id", sa.Integer(), primary_key=True),
+        sa.Column("repository_id", sa.Text(), primary_key=True),
```

Comment:
Can you clarify what this is? Is it NodeID? Full name? Better to specify explicitly in the DB.
I know there's description in the spec, but we also need the DB to be self-documented, ideally.

Reply:
So `repository_node_id`?

Reply:
I don't think we can resolve all of repo full names to NodeIDs when the request arrives.
Lets do the same as you did with commits: have a `repository_full_name` and `resolved_repo_id` or something like this.

Reply:
You are right.

Reply:
I've just realized that we have to validate the access to this repository according to the release settings. This automatically means that the repository already exists everywhere and we know the node ID. Let's leave as-is then.

Reply:
There is also a use-case when the user sends an invalid repository name (malformed, with a typo, etc.), and if we don't resolve it straight away, we accept garbage.

Reply:
> I've just realized that we have to validate the access to this repository according to the release settings.

Can you elaborate please?

> There is also a use-case when the user sends an invalid repository name (malformed, with a typo, etc.)

We can do basic validation (e.g. should look like a URL part, shouldn't be long, should start with `github.com`).

Strict validation is good, but think about different kinds of race condition that may happen on repo rename. MD fetchers may be blocked for _hours_ because of the rate limit kicking in. Thus, it's better to assume no validation or optional validation and then see how we can handle it for all the edge cases.

Reply:
If the release settings of the repo are not set to "event", we must reject the notification. This is a product feature to enforce the release settings for the account.

Renaming repositories is indeed a good example when we risk rejecting a valid release. We can document this as a product limitation and see if the clients frequently have problems with them, then re-iterate. I honestly think that this will be a rare enough problem to not sacrifice the release settings validation.

Reply:
But it might be still useful to have notifications flowing, while user plays with release settings, right?

For example, they already work with Athenian and strategy is set to `tag`. They are thinking to migrate to our Releases API, but it's not instant. They first setup everything, events start flowing, but they still want to keep old strategy or may switch back to validate something.

Reply:
Some thoughts from product:

If someone is sending correct events to mark a PR as released, can't we just ignore the settings? Why? Imagine a mono repo where someone is doing a manual release but usually PRs are released via merge to production branch. This is pretty common. 

I also prefer this strategy for large accounts who have 100s of repos, and will ask the repo owners to update their CD to include an API call but don't want to follow up with checking which repo has and which one hasn't. The user updating the deployment script is often not the Athenian user. 

Regarding strict validating, let's validate on repository and SHA, this way we can give people useful feedback if they mess either up.

Reply:
To summarize:
- Do not check if some repository has a mismatching release setting. We can return a warning in the reponse, but let's leave it for the future.
- We must resolve repository node IDs and check the security access rules like we do in the other endpoints. The product is fine with some edge cases like repository renames.

Reply:
Agree fully on your second point.

Regarding your first point, it is not a mismatching release setting, you have: Automatic, Tag, Branch, API, **but you can still use the API when you are on Automatic, Tag or Branch** because of manual individual releases or complex release processes.

Reply:
Merging the PR then :+1:

Reply:
Letting notifications flowing disregarding the release strategy is also more consistent with the other strategies. To me the equivalent sounds like considering an error when we receive a `tag creation` event on MD while the repo is configured with "branch" release strategy.

This also makes sense according to how we have split the data across the different databases: we have `persistentdata` and `metadata` dealing with "raw" data and the concept of release is "materialized" by the user settings defined in the `state` db.